internal/server/resource_network.go

```diff
@@ -2,355 +2,509 @@ package server
 
 import (
 	"context"
-	"errors"
 	"fmt"
-	"log"
 	"net"
 	"strings"
 
-	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
 
 	"github.com/hetznercloud/hcloud-go/v2/hcloud"
-	"github.com/hetznercloud/terraform-provider-hcloud/internal/network"
+	"github.com/hetznercloud/hcloud-go/v2/hcloud/exp/kit/sliceutil"
 	"github.com/hetznercloud/terraform-provider-hcloud/internal/util"
 	"github.com/hetznercloud/terraform-provider-hcloud/internal/util/control"
 	"github.com/hetznercloud/terraform-provider-hcloud/internal/util/hcloudutil"
-	"github.com/hetznercloud/terraform-provider-hcloud/internal/util/merge"
 )
 
-// NetworkResourceType is the type name of the Hetzner Cloud Server
-// network resource.
 const NetworkResourceType = "hcloud_server_network"
 
-// NetworkResource creates a Terraform schema for the hcloud_server_network
-// resource.
-func NetworkResource() *schema.Resource {
-	return &schema.Resource{
-		CreateContext: resourceServerNetworkCreate,
-		ReadContext:   resourceServerNetworkRead,
-		UpdateContext: resourceServerNetworkUpdate,
-		DeleteContext: resourceServerNetworkDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: schema.ImportStatePassthroughContext,
-		},
-		Schema: map[string]*schema.Schema{
-			"network_id": {
-				Type:     schema.TypeInt,
-				Optional: true,
-				ForceNew: true,
+var _ resource.Resource = (*networkResourceImpl)(nil)
+var _ resource.ResourceWithConfigure = (*networkResourceImpl)(nil)
+var _ resource.ResourceWithConfigValidators = (*networkResourceImpl)(nil)
+var _ resource.ResourceWithImportState = (*networkResourceImpl)(nil)
+
+type networkResourceImpl struct {
```

Comment:
Is there a reason this resource uses a different naming schema to other plugin framework resources?

```suggestion
type NetworkResource struct {
```

Reply:
It is following the ssh key resource naming pattern. Where we could not use the `resource` name since it is a package.

But we could use `Resource` for sure.